cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -400,6 +412,135 @@ async function activateOption(option, loadBalancerDNS, loadBalancerHostedZone) {
     }).promise();
 }
 
+/**
+ * Retrieve all certificates used for the service and cache the results.
+ * @returns {Array<Object>} An array of descriptions for the certificates used by the service.
+ */
+async function serviceCertificates() {
+    if (cachedServiceCertificates) {
+        return cachedServiceCertificates;
+    }
+    let { ResourceTagMappingList } = await clients.resourceGroupsTaggingAPI().getResources({
+        TagFilters: [
+            {
+                Key: "copilot-application",
+                Values: [appName],
+            },
+            {
+                Key: "copilot-environment",
+                Values: [envName],
+            },
+            {
+                Key: "copilot-service",
+                Values: [serviceName],
+            }
+        ],
+        ResourceTypeFilters: ["acm:certificate"]
+    }).promise();
+
+    cachedServiceCertificates = [];
+    let promises = [];
+    for (let {ResourceARN: arn} of ResourceTagMappingList) {
+        let promise = clients.acm().describeCertificate({
+            CertificateArn: arn
+        }).promise().then(( { Certificate} ) => {
+            cachedServiceCertificates.push(Certificate);
+        });
+        promises.push(promise);
+    }
+    await Promise.all(promises);
+    return cachedServiceCertificates;
+}
+
+/**
+ * Retrieve validation options that will be unused by the service after deleting the certificate.
+ * An option is deemed pending unused if it validates only the certificate(s) pending deletion.
+ *
+ * @param {Set<String>} aliases is the subject alternative names of the certificate(s) pending deletion.
+ * @returns {Set<Object>} The options that will be unused by the service.
+ */
+async function optionsUnusedByService(aliases) {
+    // NOTE: `Set` uses object reference to identify elements. That is, two options with the same value will not be
+    // identify as the same option. Therefore, we use map with object string as keys to identify object with actual
+    // values instead of object reference.
+    let optionsInUse = new Map();
+    let optionsPendingUnusedByService = new Map();
+
+    let certificates = await serviceCertificates();
+    for (const {DomainName: domainName, SubjectAlternativeNames: sans, DomainValidationOptions: validationOptions} of certificates) {
+        if (domainName === certificateDomain && setEqual(aliases.add(certificateDomain), new Set(sans))) {
```

Comment:
I'm struggling to understand what `setEqual(aliases.add(certificateDomain), new Set(sans))`does 🤔 

1. Why do we need to add `certificateDomain` to `aliases`?
2. Do we want to constantly append to `aliases`? or should we create a copy of the set before appending

Reply:
1. Because the "[SubjectAlternativeName](https://docs.aws.amazon.com/acm/latest/APIReference/API_CertificateDetail.html#ACM-Type-CertificateDetail-SubjectAlternativeNames)" field returned from the API always include the canonical domain name of the certificate.
2. Good point! We should create a copy before appending.